### PR TITLE
nixos-rebuild: Add option for building and/or deploying on a remote host

### PR DESCRIPTION
**Pull request by rickynils**
nixos-rebuild: Add option for building and/or deploying on a remote host

This commit adds the options --build-host and --target-host to nixos-rebuild.

--build-host instructs nixos-rebuild to perform all nix builds on the 
specified host (via ssh). Build results are then copied back to the 
local machine and used when activating the system.

--build-target instructs nixos-rebuild to activate the configuration
not on the local machine but on the specified remote host. Build
results are copied to the target machine and then activated there (via ssh).

It is possible to combine the usage of --build-host and --target-host,
in which case you can perform the build on one remote machine and deploy
the configuration to another remote machine. The only requirement is that
the build host has a working ssh connection to the target host (if the 
target is not local), and that the local machine can connect to both
the target and the build host. Also, your user must be allowed to copy
nix closures between the local machine and the target and host machines.

At no point in time are the configuration sources (the nix files) copied
anywhere. Instead, nix evaluation always happens locally
(with nix-instantiate). The drv-file is then copied and realised remotely
(with nix-store).

As a convenience, if only --target-host is specified, --build-host is
implicitly set to that host too. So if you want to build locally and deploy
remotely you have to explicitly set "--build-host localhost".

To activate (test, boot or switch) you need to have root access to the 
target host. You can specify this by "--target-host root@myhost".

I have tested the obvious scenarios and they are working. Some of the 
combinations of --build-host and --target-host and the various actions might
not make much sense, and should maybe be forbidden (like setting a remote
target host when building a VM), and some combinations might not work at all.

### CODE REVIEW

**Comment by rickynils** [comment]
@edolstra @shlevy @domenkozar Do we want do have this kind of functionality in `nixos-rebuild`, or is it better suited for a separate tool?

Pros: Makes it very convenient to deploy NixOS machines that you for some reason don't want to manage with nixops or similar. Also, this is not a large commit at all. If you were to do this in a separate tool you would have to copy most of the stuff from `nixos-rebuild` to that tool anyway.

If we think this is a good idea, some feedback on the ui/semantics/implementation is welcome. Documentation will be written when that is settled.

@proger This basically replicates the remote build functionality from [upcast](https://github.com/zalora/upcast), but in a more NixOS-centric fashion. Do you have any input on the user interface described in the commit message?

**Comment by shlevy** [comment]
Nix already has the remote build hook, why do we need `--build-host`? I like `--target-host` though

**Comment by shlevy** [comment]
Ah, if you have both build-host and target-host then nothing is copied back to the instantiating machine. Cool. :+1: for me for the concept, haven't done a thorough code review yet

**Comment by rickynils** [comment]
@shlevy Yeah, the remote build hook always copies back to the local machine which is pointless if the target is remote. Also, if everything already is built remotely, but not locally, nix will start building everything locally or possibly on another remote that doesn't have them, while in the end nothing will be copied to the target (this can be somewhat mitigated by having the remote machine as an ssh-binary-cache, though).

**Comment by domenkozar** [comment]
:+1: for the idea. This superseeds nixops `none` backend for my needs.

**Comment by rickynils** [comment]
To test this, build `nixos-rebuild` like this (`configuration.nix` can be an empty config, it is not really used):

```
nix-build '<nixpkgs/nixos>' \
  -I nixpkgs=https://github.com/rickynils/nixpkgs/archive/nixos-rebuild-remote.tar.gz \
  -I nixos-config=./configuration.nix \
  -A config.system.build.nixos-rebuild
```

Then (since `nixos-rebuild` re-execs itself we need to use my branch also when building NixOS):

```
./result/bin/nixos-rebuild \
  -I nixpkgs=https://github.com/rickynils/nixpkgs/archive/nixos-rebuild-remote.tar.gz \
  ...
```

**Comment by proger** [comment]
@rickynils looks neat! As a side comment, it'd be really nice to see an implementation for nix-instantiate+nix-copy-closure using [`cmdBuildDerivation`](https://github.com/NixOS/nix/commit/1511aa9f488ba0762c2da0bf8ab61b5fde47305d) like [hydra does](https://github.com/NixOS/hydra/commit/5370be9f52ada02b99ada6e92bb188b25de618f1) — since around nix 1.9 nix-copy-closure for a derivation started to transfer a bunch of extra stuff and `cmdBuildDerivation` seems to address a related problem in a cuter way.

You should probably point out somewhere that using import-from-derivation kills all the remote building fun.

**Comment by proger** [comment]
> The only requirement is that the build host has a working ssh connection to the target host.

> So if you want to build remotely and deploy locally you have to set "--targetHost localhost".

These two seem mutually exclusive

**Comment by proger** [comment]
> Nix already has the remote build hook, why do we need --build-host?

The main problem with the remote build hook is doing remote deployments unnecessarily transfers built store paths back and forth to the client machine. ("[starbucks-mode](https://github.com/zalora/upcast#remote-builds)")

**Comment by rickynils** [comment]
@proger Well-spotted, that is mutually exclusive, and the implementation didn't support it either. I've fixed it so that builds are pulled from the build host if the target is local now.

The hydra approach seems nice, thanks for bringing it to my attention.

**Comment by rickynils** [comment]
I've switched the implicit dependency from `--build-host ==> --target-host` to `--target-host ==> --build-host`. In other words, if `--build-host` is not set, it will get the value of `--target-host`. I think that feels more natural.

Also, build results are now always copied to the target. So if you do `nixos-rebuild --build-host myhost build` the build results will end up on the local machine in the end. The behavior is in line with how a local-only build behaves (but that particular use case is ofc already served today by the remote build hook).

File: nixos/modules/installer/tools/nixos-rebuild.sh [hunk 1]
```diff
@@ -80,6 +90,87 @@ while [ "$#" -gt 0 ]; do
...
+        *)
+          buildArgs+=("$i")
+          ;;
+      esac
+    done
+
+    local drv="$(nix-instantiate "${instArgs[@]}" "${extraBuildFlags[@]}")"
+    if [ -a "$drv" ]; then
+      nix-copy-closure --to "$buildHost" "$drv"
+      ssh "$buildHost" nix-store -r "$drv" "${buildArgs[@]}"
```
Comment (nbp) [inline nixos/modules/installer/tools/nixos-rebuild.sh]: I will recommend to use a ssh master connection, such that the user does not have to wonder why he has to login twice.

File: nixos/modules/installer/tools/nixos-rebuild.sh [hunk 2]
```diff
@@ -200,31 +291,38 @@ fi
...
-        pathToConfig=./result
+        pathToConfig="$(nixBuild '<nixpkgs/nixos>' -A system -k "${extraBuildFlags[@]}")"
     elif [ "$action" = build-vm ]; then
-        nix-build '<nixpkgs/nixos>' -A vm -k "${extraBuildFlags[@]}" > /dev/null
-        pathToConfig=./result
+        pathToConfig="$(nixBuild '<nixpkgs/nixos>' -A vm -k "${extraBuildFlags[@]}")"
     elif [ "$action" = build-vm-with-bootloader ]; then
-        nix-build '<nixpkgs/nixos>' -A vmWithBootLoader -k "${extraBuildFlags[@]}" > /dev/null
-        pathToConfig=./result
+        pathToConfig="$('<nixpkgs/nixos>' -A vmWithBootLoader -k "${extraBuildFlags[@]}")"
```
Comment (nbp) [inline nixos/modules/installer/tools/nixos-rebuild.sh]: This lacks the `nixBuild` function.

File: nixos/modules/installer/tools/nixos-rebuild.sh [hunk 3]
```diff
@@ -164,16 +255,16 @@ if [ -n "$buildNix" ]; then
...
                     echo "warning: don't know how to get latest Nix" >&2
                 fi
                 # Older version of nix-store -r don't support --add-root.
-                [ -e $tmpDir/nix ] || ln -sf $nixStorePath $tmpDir/nix
+                buildHostCmd [ -e $tmpDir/nix ] || ln -sf $nixStorePath $tmpDir/nix
             fi
         fi
     fi
-    PATH=$tmpDir/nix/bin:$PATH
+    PATH="$tmpDir/nix/bin:$PATH"
```
Comment (nbp) [inline nixos/modules/installer/tools/nixos-rebuild.sh]: The whole point of evaluating the Nix interpreter before the configuration, is to ensure that we have the lastest version for instantiating and building.  If we specify a build host, then we probably want to use an update-to-date version of Nix as well for evaluating the configuration.  In this particular case, having a build host set would set the path which does not exists locally.

File: nixos/modules/installer/tools/nixos-rebuild.sh [hunk 4]
```diff
@@ -152,9 +243,9 @@ fi
 
 if [ -n "$buildNix" ]; then
     echo "building Nix..." >&2
-    if ! nix-build '<nixpkgs/nixos>' -A config.nix.package -o $tmpDir/nix "${extraBuildFlags[@]}" > /dev/null; then
-        if ! nix-build '<nixpkgs/nixos>' -A nixFallback -o $tmpDir/nix "${extraBuildFlags[@]}" > /dev/null; then
-            if ! nix-build '<nixpkgs>' -A nix -o $tmpDir/nix "${extraBuildFlags[@]}" > /dev/null; then
+    if ! nixBuild '<nixpkgs/nixos>' -A config.nix.package -o $tmpDir/nix "${extraBuildFlags[@]}" > /dev/null; then
```
Comment (nbp) [inline nixos/modules/installer/tools/nixos-rebuild.sh]: I am not sure, but I think I might prefer if we were to use `nix-instantiate` and `nix-store -r` explicitly here.
Thus this would just be a matter of prefixing all the commands with the location of the action.

```
  if ! drv=$(nix-instantiate …) ; then
    …
  fi
  if test -n "$drv"; then
    buildHostCmd nix-store -r … ;
  fi
```

**Comment by nbp** [comment]
I really like the idea, as this solve most of my issues where I only manage a few computers manually.

I still wonder how this would work with automatic update features.  I guess we should side-load a module which raise an assertion if the auto-update feature is enabled at the same time.

**Comment by rickynils** [comment]
@nbp Thank you for the feedback! The nix version discrepancy between instantiate/realise had slipped me by, I'll try to fix that. And of course the other, more obvious bugs you pointed out.

Yeah, my use case is just that, those machines that are not really part of a cluster/network of machines, and that needs to be managed independently but still in a nice centralised fashion.

The automatic updates, are these the security updates? How is that handled for ordinary, local, `nixos-rebuild`?

**Comment by nbp** [comment]
> The automatic updates, are these the security updates? How is that handled for ordinary, local, `nixos-rebuild`?

Yes, I think this is some sort of cron jobs which does a `nixos-rebuild` command.  I guess the problem also exists for any instances of `nixops`.

**Comment by the-kenny** [comment]
This is great! Any updates on this?

**Comment by rickynils** [comment]
I have not forgotten about this, just busy with other things. I will fix the outstanding issues as soon as possible.

**Comment by rickynils** [comment]
@nbp I've fixed the Nix building parts now, I hope. Things got a bit messier since we need to build nix _both_ locally and remotely.

I've also made ssh use a master connection (and `NIX_SSHOPTS`)

**Comment by rickynils** [comment]
Easiest way to test this:

Single-line install with `nix-env`:
```
nix-env -iA config.system.build.nixos-rebuild --arg configuration '{}' -I nixpkgs=https://github.com/rickynils/nixpkgs/archive/nixos-rebuild-remote.tar.gz -f '<nixpkgs/nixos>'
```

The run with `_NIXOS_REBUILD_REEXEC=1` to keep `nixos-rebuild` from re-executing with the the version in `NIX_PATH`:

```
_NIXOS_REBUILD_REEXEC=1 nixos-rebuild ...
```

This way, you can try this out with any existing NixOS configuration.

**Comment by edolstra** [comment]
IMHO this duplicates functionality already provided by NixOps, so I'm not too keen on complicating `nixos-rebuild`...

**Comment by rickynils** [comment]
@edolstra I agree there is some duplication between this, the remote build hook and NixOps. But...

The build hook supports remote builds but forces everything to be copied back to the local machine.

NixOps doesn't support remote builds except through the build hook (or does it?). Also, NixOps comes with other stuff: the network definition and the stateful deployment model. There are situations where that is not needed or not acceptable. NixOps is a nice tool that has a particular way of handling NixOS deployments, but I don't think that has to be the only way to deploy NixOS on remote hosts.

`nixos-rebuild` handles NixOS builds and activation. The ssh functionality extends that to remote hosts. I don't see that as a terrible feature creep.

I think that, ideally, the ordinary `nix-build` tools should support remote-only builds. `nixos-rebuild` could then be a bit simpler (but would still need the concepts of target and build hosts).

We could also have a separate `nixos-rebuild-remote` tool. That is OK with me, but it would probably not be able to use `nixos-rebuild`, but rather re-implement it (like I've done in this PR). This means we have two places to maintain the NixOS activation process in.

To summarize: If people don't regard this tool as useful (because of the existence of NixOps or whatever), then I can move it outside nixpkgs. If people find it useful but don't want to bloat `nixos-rebuild`, I can move it to a separate (`nixos-rebuild-remote`?) tool. My opinion is that it is useful (I use it, and have used variants of it for a long time), and that the combination of `nixos-rebuild` + `nixos-rebuild-remote` would be less maintainable than this implementation of `nixos-rebuild`.

**Comment by waern** [comment]
This would be useful to me as a lightweight deployment method for smaller things running on a single remote machine.

**Comment by edolstra** [comment]
@rickynils Okay. Can you update the nixos-rebuild manpage and fix the indentation (4 chars per level).

**Comment by garbas** [comment]
:beers:

**Comment by rickynils** [comment]
@edolstra Indentation fixed and man page updated

**Comment by aborsu** [comment]
:+1: This would be a nice to have.

**Comment by rickynils** [comment]
@edolstra OK to merge now?
